src/supervisor/message_handlers.py

```diff
@@ -0,0 +1,190 @@
+"""
+message_handlers.py: Holds all the different message handlers
+"""
+
+import gzip
+import io
+import logging
+from abc import ABC, abstractmethod
+from email.mime.application import MIMEApplication
+from email.mime.multipart import MIMEMultipart
+from email.mime.text import MIMEText
+from pathlib import Path
+from smtplib import SMTP
+
+import pkg_resources
+
+
+class MessageHandler(ABC):  # pylint: disable=too-few-public-methods
+    """Base class for all message handlers.
+
+    Methods
+    -------
+    send_message(message_details)
+        Parse a MessageDetails object and send a message using handler-specific logic
+    """
+
+    @abstractmethod
+    def send_message(self, message_details):
+        """Send a notification using the target-specific logic (email, slack, etc)
+
+        Parameters
+        ----------
+        message_details : MessageDetails
+            The data to be sent in the notification
+        """
+
+
+class EmailHandler(MessageHandler):  # pylint: disable=too-few-public-methods
+    """Send a notification via email
+
+    Attributes
+    ----------
+    email_settings : dict
+        From and To addresses, SMTP Server and Port
+
+    Methods
+    -------
+    send_message(message_details)
+        Build a message, create an SMTP object, and send the message
+    _build_message(message_details)
+        Create email to be sent as a MIMEMultipart object
+    _build_gzip_attachment(input_path)
+        gzip input_path into a MIMEApplication object
+    """
```

Comment:
**question**: This doc format is nice. Where did you find it? Is it a standard defined somewhere? I'm not sure that I like having to remember to update the parameters in two locations but the popups in VS Code look nice.

Reply:
It's a bit verbose, but I thought it was worth a shot. It's the numpy docstring standard (https://numpydoc.readthedocs.io/en/latest/format.html) that I stumbled on while perusing https://realpython.com/documenting-python-code/#docstring-formats. It may need some tweaking to make sure it formats properly in VSCcode, pycharm, and whatever else people are using in the office.